helix-lsp/src/lib.rs

```diff
@@ -649,10 +649,25 @@ impl Notification {
     }
 }
 
+#[derive(Debug)]
+enum Clients {
+    /// A set of running clients
+    Clients(Vec<Arc<Client>>),
+    /// A marker that clients for a language server were manually stopped
+    /// and should not be automatically restarted.
+    Tombstone,
+}
+
+impl Clients {
+    fn new(client: Arc<Client>) -> Self {
+        Self::Clients(vec![client])
+    }
+}
+
 #[derive(Debug)]
 pub struct Registry {
     inner: SlotMap<LanguageServerId, Arc<Client>>,
-    inner_by_name: HashMap<LanguageServerName, Vec<Arc<Client>>>,
+    inner_by_name: HashMap<LanguageServerName, Clients>,
```

Comment:
I wonder if we really need a new enum. Ii think it would be sufficient to have the following:

* not in hashmap: no lsp started yet.
* in hashmap and vec length >= 1: lsp is active (either add a new client or reuse an existing one)
* in hashmap and vec length == 0: We can only endup here if `:lsp-stop` is called, don't start a LSP instance for this document.

This would not require adding a new type and would really only require two changes:
* leave behind an empty list in the `stop` function
* `Registry::get`: don't start the LSP if the list is empty

it's a bit of a hack but it would be a much smaller change.

Reply:
Good point, it's a much smaller diff. It's all an implementation detail of Registry and pretty well contained there so the enum is probably not necessary - we can replace it with some comments explaining the special value